shell/browser/api/atom_api_url_request_ns.cc

```diff
@@ -0,0 +1,536 @@
+  if (!request_)
+    return false;
+  if (!net::HttpUtil::IsValidHeaderName(name))
+    return false;
+  if (!net::HttpUtil::IsValidHeaderValue(value))
+    return false;
+  request_->headers.SetHeader(name, value);
+  return true;
+}
+
+void URLRequestNS::RemoveExtraHeader(const std::string& name) {
+  if (request_)
+    request_->headers.RemoveHeader(name);
+}
+
+void URLRequestNS::SetChunkedUpload(bool is_chunked_upload) {
+  if (request_)
+    is_chunked_upload_ = is_chunked_upload;
+}
+
+mate::Dictionary URLRequestNS::GetUploadProgress() {
+  mate::Dictionary progress = mate::Dictionary::CreateEmpty(isolate());
+  if (loader_) {
+    if (request_)
+      progress.Set("started", false);
+    else
+      progress.Set("started", true);
+    progress.Set("current", upload_position_);
+    progress.Set("total", upload_total_);
+    progress.Set("active", true);
+  } else {
+    progress.Set("active", false);
+  }
+  return progress;
+}
+
+int URLRequestNS::StatusCode() const {
+  if (response_headers_)
+    return response_headers_->response_code();
+  return -1;
+}
+
+std::string URLRequestNS::StatusMessage() const {
+  if (response_headers_)
+    return response_headers_->GetStatusText();
+  return "";
+}
+
+net::HttpResponseHeaders* URLRequestNS::RawResponseHeaders() const {
+  return response_headers_.get();
+}
+
+uint32_t URLRequestNS::ResponseHttpVersionMajor() const {
+  if (response_headers_)
+    return response_headers_->GetHttpVersion().major_value();
+  return 0;
+}
+
+uint32_t URLRequestNS::ResponseHttpVersionMinor() const {
+  if (response_headers_)
+    return response_headers_->GetHttpVersion().minor_value();
+  return 0;
+}
+
+void URLRequestNS::OnDataReceived(base::StringPiece data,
+                                  base::OnceClosure resume) {
+  // In case we received an unexpected event from Chromium net, don't emit any
+  // data event after request cancel/error/close.
+  if (!(request_state_ & STATE_ERROR) && !(response_state_ & STATE_ERROR)) {
+    v8::HandleScope handle_scope(isolate());
+    v8::Local<v8::Value> buffer;
+    auto maybe = node::Buffer::Copy(isolate(), data.data(), data.size());
+    if (maybe.ToLocal(&buffer))
+      Emit("data", buffer);
+  }
+  std::move(resume).Run();
+}
+
+void URLRequestNS::OnRetry(base::OnceClosure start_retry) {}
+
+void URLRequestNS::OnComplete(bool success) {
+  if (success) {
+    // In case we received an unexpected event from Chromium net, don't emit any
+    // data event after request cancel/error/close.
+    if (!(request_state_ & STATE_ERROR) && !(response_state_ & STATE_ERROR)) {
+      response_state_ |= STATE_FINISHED;
+      Emit("end");
+    }
+  } else {  // failed
+    // Only emit error when there is no previous failure.
+    // This is to align with the behavior of non-NetworkService implementation.
+    if (!(request_state_ & STATE_FAILED))
+      EmitRequestError(net::ErrorToString(loader_->NetError()));
```

Comment:
Should `response_state_` be set to `STATE_FAILED` in the `else` block ? At the moment I see that `response_state_` only toggles between `STATE_STARTED` and `STATE_FINISHED` , there is no indication of `STATE_ERROR` but we do check for this in `Cancel` and other places.

Reply:
Nice catch. The `SimpleURLLoader` does not distinguish between response errors and request errors, so I have pushed a change to emit error depending on current state.